flows/inference.py

```diff
@@ -118,6 +122,111 @@ def to_json(self) -> dict:
+    unknown_failures: list[BaseException] = []
+    classifier_name: str
+    classifier_alias: str
+
+    @property
+    def failed(self) -> bool:
+        """Whether the batch failed, True if failed."""
+
+        return self.failed_document_stems != [] or self.unknown_failures != []
+
+
+class InferenceException(Exception):
+    """
+    Exception raised when inference fails.
+
+    The data attribute of the prefect State when type is FAILED must be an object that
+    an exception can be raised from. Thus, we declare a custom exception that can also
+    be used to transfer the results of the all the batch inference runs.
+    """
+
+    def __init__(self, message: str, data: dict[str, Any]):
+        super().__init__(message)
+
+        self.message = message
+        self.data = data
+
+
+class InferenceResult(BaseModel):
+    """Result from running inference on all batches of documents."""
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    batch_inference_results: list[BatchInferenceResult] = []
+    unexpected_failures: list[BaseException | FlowRun] = []
+    successful_classifier_specs: list[ClassifierSpec] = []
+    failed_classifier_specs: list[ClassifierSpec] = []
+
+    @property
+    def failed(self) -> bool:
+        """Whether the inference failed, True if failed."""
+
+        return (
+            any([result.failed for result in self.batch_inference_results])
+            or self.unexpected_failures != []
+        )
+
+    @cached_property
+    def successful_document_stems(self) -> set[DocumentStem]:
+        """
+        The set of document stems that were successfully processed.
+
+        A document stem is considered successful if it was succesful across all classifiers. For example,
+        if a document successfully had inference run in one batch for classifier A, but failed for classifier B,
+        then the document stem is considered unsuccessful.
+
+        This is as the document would fail aggregation if there was a missing inference result for a classifier.
+        """
+
+        failed_document_stems = set(
+            document_stem
+            for batch_inference_result in self.batch_inference_results
+            for document_stem, _ in batch_inference_result.failed_document_stems
+        )
+
+        return set(
+            document_stem
+            for batch_inference_result in self.batch_inference_results
+            for document_stem in batch_inference_result.successful_document_stems
+            if document_stem not in failed_document_stems
+        )
+
+    @cached_property
+    def failed_document_stems(self) -> set[DocumentStem]:
+        """The set of document stems that failed to be processed."""
+
+        return set(
+            document_stem
+            for batch_inference_result in self.batch_inference_results
+            for document_stem, _ in batch_inference_result.failed_document_stems
+        )
```

Comment:
This class could be simplified by removing the failed_document_stems method as well as the classifier fields as these aren't currently used (were added and then the code refactored). They may prove useful soon though?